In [1]:
import warnings 
warnings.filterwarnings('ignore')

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np


In [3]:
df = pd.read_csv('smoking_driking_dataset_Ver01.csv')
df.head()

sex  age  height  weight  waistline  sight_left  sight_right  hear_left  \
0  Male   35     170      75       90.0         1.0          1.0        1.0   
1  Male   30     180      80       89.0         0.9          1.2        1.0   
2  Male   40     165      75       91.0         1.2          1.5        1.0   
3  Male   50     175      80       91.0         1.5          1.2        1.0   
4  Male   50     165      60       80.0         1.0          1.2        1.0   

   hear_right    SBP  ...  LDL_chole  triglyceride  hemoglobin  urine_protein  \
0         1.0  120.0  ...      126.0          92.0        17.1            1.0   
1         1.0  130.0  ...      148.0         121.0        15.8            1.0   
2         1.0  120.0  ...       74.0         104.0        15.8            1.0   
3         1.0  145.0  ...      104.0         106.0        17.6            1.0   
4         1.0  138.0  ...      117.0         104.0        13.8            1.0   

   serum_creatinine  SGOT_AST  SGOT_ALT  gamma_GTP  SMK_stat_type_cd  DRK_YN  
0               1.0      21.0      35.0       40.0               1.0       Y  
1               0.9      20.0      36.0       27.0               3.0       N  
2               0.9      47.0      32.0       68.0               1.0       N  
3               1.1      29.0      34.0       18.0               1.0       N  
4               0.8      19.0      12.0       25.0               1.0       N  

[5 rows x 24 columns]

In [4]:
pd.set_option('display.max_columns',None)
df.head()

sex  age  height  weight  waistline  sight_left  sight_right  hear_left  \
0  Male   35     170      75       90.0         1.0          1.0        1.0   
1  Male   30     180      80       89.0         0.9          1.2        1.0   
2  Male   40     165      75       91.0         1.2          1.5        1.0   
3  Male   50     175      80       91.0         1.5          1.2        1.0   
4  Male   50     165      60       80.0         1.0          1.2        1.0   

   hear_right    SBP   DBP   BLDS  tot_chole  HDL_chole  LDL_chole  \
0         1.0  120.0  80.0   99.0      193.0       48.0      126.0   
1         1.0  130.0  82.0  106.0      228.0       55.0      148.0   
2         1.0  120.0  70.0   98.0      136.0       41.0       74.0   
3         1.0  145.0  87.0   95.0      201.0       76.0      104.0   
4         1.0  138.0  82.0  101.0      199.0       61.0      117.0   

   triglyceride  hemoglobin  urine_protein  serum_creatinine  SGOT_AST  \
0          92.0        17.1            1.0               1.0      21.0   
1         121.0        15.8            1.0               0.9      20.0   
2         104.0        15.8            1.0               0.9      47.0   
3         106.0        17.6            1.0               1.1      29.0   
4         104.0        13.8            1.0               0.8      19.0   

   SGOT_ALT  gamma_GTP  SMK_stat_type_cd DRK_YN  
0      35.0       40.0               1.0      Y  
1      36.0       27.0               3.0      N  
2      32.0       68.0               1.0      N  
3      34.0       18.0               1.0      N  
4      12.0       25.0               1.0      N

In [5]:
df.isnull().sum()

sex                 0
age                 0
height              0
weight              0
waistline           0
sight_left          0
sight_right         0
hear_left           0
hear_right          0
SBP                 0
DBP                 0
BLDS                0
tot_chole           0
HDL_chole           0
LDL_chole           0
triglyceride        0
hemoglobin          0
urine_protein       0
serum_creatinine    0
SGOT_AST            0
SGOT_ALT            0
gamma_GTP           0
SMK_stat_type_cd    0
DRK_YN              0
dtype: int64

In [6]:
X = df.drop('DRK_YN', axis=1)
y = df['DRK_YN']

In [7]:
from sklearn.model_selection import train_test_split

xtrain, xtest, ytrain, ytest = train_test_split(X,y, test_size=0.3, random_state=42)

In [8]:
xtrain.shape

(693942, 23)

In [9]:
ytrain.shape

(693942,)

In [10]:
xtest.shape

(297404, 23)

In [11]:
ytest.shape

(297404,)

Lebel Encoding

In [12]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [13]:
categorical_columns = X.select_dtypes(include=['object']).columns

for col in categorical_columns:
    xtrain[col] = le.fit_transform(xtrain[col])
    xtest[col] = le.fit_transform(xtest[col])
    

In [14]:
xtrain.head()

sex  age  height  weight  waistline  sight_left  sight_right  \
751370    0   55     155      60       93.0         0.7          0.9   
823097    0   65     150      55       82.0         0.8          0.7   
439934    0   50     165      60       78.4         1.2          1.0   
242598    1   65     160      55       86.0         0.8          0.8   
346295    0   40     160      60       69.0         1.5          1.5   

        hear_left  hear_right    SBP   DBP   BLDS  tot_chole  HDL_chole  \
751370        1.0         1.0  110.0  76.0  111.0      185.0       52.0   
823097        1.0         1.0  130.0  80.0   95.0      180.0       51.0   
439934        1.0         1.0  115.0  61.0   86.0      185.0       76.0   
242598        1.0         1.0  120.0  80.0  118.0      150.0       35.0   
346295        1.0         1.0  130.0  80.0   96.0      188.0       73.0   

        LDL_chole  triglyceride  hemoglobin  urine_protein  serum_creatinine  \
751370      115.0          90.0        13.2            1.0               1.0   
823097      108.0         101.0        11.6            1.0               0.8   
439934       88.0         103.0        12.2            1.0               0.7   
242598       81.0         166.0        11.7            1.0               1.0   
346295       98.0          84.0        10.8            1.0               0.6   

        SGOT_AST  SGOT_ALT  gamma_GTP  SMK_stat_type_cd  
751370      34.0      62.0       51.0               2.0  
823097      28.0      19.0       18.0               1.0  
439934      26.0      25.0       33.0               1.0  
242598      27.0      21.0       17.0               3.0  
346295      20.0       9.0       10.0               1.0

In [15]:
xtrain.shape

(693942, 23)

In [16]:
xtest.shape

(297404, 23)

In [17]:
xtest.head()

sex  age  height  weight  waistline  sight_left  sight_right  \
676014    0   45     160      60       85.0         1.2          1.0   
958905    1   30     175      70       87.0         0.9          0.9   
472540    1   35     180      85       93.0         1.2          1.5   
244194    1   35     175      80       89.0         1.5          1.5   
88447     0   40     160      70       97.0         1.5          1.0   

        hear_left  hear_right    SBP   DBP   BLDS  tot_chole  HDL_chole  \
676014        1.0         1.0  113.0  77.0   84.0      195.0       82.0   
958905        1.0         1.0  135.0  78.0   90.0      223.0       50.0   
472540        1.0         1.0  129.0  86.0  102.0      206.0       46.0   
244194        1.0         1.0  119.0  76.0   77.0      202.0       62.0   
88447         1.0         1.0  124.0  67.0   78.0      166.0       58.0   

        LDL_chole  triglyceride  hemoglobin  urine_protein  serum_creatinine  \
676014      100.0          64.0        13.3            1.0               0.8   
958905      147.0         127.0        15.9            2.0               1.2   
472540      119.0         204.0        16.6            1.0               1.0   
244194      130.0          50.0        15.2            1.0               0.9   
88447        91.0          81.0        12.8            1.0               0.5   

        SGOT_AST  SGOT_ALT  gamma_GTP  SMK_stat_type_cd  
676014      16.0      13.0       14.0               1.0  
958905      24.0      28.0       16.0               1.0  
472540      34.0      61.0       25.0               2.0  
244194      18.0      17.0       17.0               1.0  
88447       15.0      17.0       14.0               1.0

# 1. The purpose of this dataset is to classify Drinker or Not.

In [18]:
from sklearn.tree import DecisionTreeClassifier

dtc = DecisionTreeClassifier(random_state=42)

dtc.fit(xtrain,ytrain)

DecisionTreeClassifier(random_state=42)

In [19]:
predict_train_data = dtc.predict(xtrain)
predict_train_data

array(['N', 'N', 'N', ..., 'Y', 'N', 'Y'], dtype=object)

In [20]:
ytrain

751370    N
823097    N
439934    N
242598    Y
346295    N
         ..
259178    Y
365838    N
131932    Y
671155    N
121958    Y
Name: DRK_YN, Length: 693942, dtype: object

In [21]:
predict_test_data = dtc.predict(xtest)
predict_test_data

array(['Y', 'Y', 'Y', ..., 'N', 'N', 'Y'], dtype=object)

In [22]:
ytest

676014    N
958905    N
472540    N
244194    Y
88447     N
         ..
514545    Y
867505    Y
555098    Y
751529    Y
160158    N
Name: DRK_YN, Length: 297404, dtype: object

# Perform Decision Tree Classifier & KNN. Evaluate Both ML Models by Confusion Matrix and Relevant Performance Parameters

In [23]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, roc_auc_score, roc_curve, auc, balanced_accuracy_score


In [24]:
train_conf_matrix = confusion_matrix(ytrain, predict_train_data)
train_conf_matrix

array([[346711,      0],
       [     0, 347231]], dtype=int64)

In [25]:
test_conf_matrix = confusion_matrix(ytest, predict_test_data)
test_conf_matrix

array([[96158, 52989],
       [52453, 95804]], dtype=int64)

In [26]:
train_accuracy = accuracy_score(ytrain, predict_train_data)
train_accuracy

1.0

In [27]:
test_accuracy = accuracy_score(ytest, predict_test_data)
test_accuracy

0.6454587026401797

In [28]:
train_f1_score = f1_score(ytrain, predict_train_data, pos_label='Y')
train_f1_score

1.0

In [29]:
test_f1_score = f1_score(ytest, predict_test_data, pos_label='Y')
test_f1_score

0.6450361891937385

In [30]:
train_precision_score = precision_score(ytrain, predict_train_data, pos_label='Y')
train_precision_score

1.0

In [31]:
test_precision_score = precision_score(ytest, predict_test_data, pos_label='Y')
test_precision_score

0.6438743758106901

In [32]:
train_recall_score = recall_score(ytrain, predict_train_data, pos_label='Y')
train_recall_score

1.0

In [33]:
test_recall_score = recall_score(ytest, predict_test_data, pos_label='Y')
test_recall_score

0.6462022029313962

In [34]:
train_balanced_accuracy_score = balanced_accuracy_score(ytrain, (predict_train_data=='Y').astype(str))
train_balanced_accuracy_score

0.0

In [35]:
test_balanced_accuracy_score = balanced_accuracy_score(ytest, (predict_test_data=='Y').astype(str))
test_balanced_accuracy_score

0.0

# Sir my Question is 'why balanced_accuracy_score is comes zero' How can I improve balanced accuracy ?

KNN

In [36]:
from sklearn.neighbors import KNeighborsClassifier

knn_model = KNeighborsClassifier(n_neighbors=833)


knn_model.fit(xtrain, ytrain)

Rule of Thumb: A common starting point is to set n_neighbors to the square root of the number of samples in your dataset. In your case, with 693,942 samples, you might consider a value around 832 (since √693,942 ≈ 832)

In [37]:
knn_model.fit(xtest, ytest)

KNeighborsClassifier(n_neighbors=833)

In [38]:
knn_model.fit(xtrain, ytrain)

KNeighborsClassifier(n_neighbors=833)

In [39]:
xtrain = np.arange(0, np.prod(xtrain.shape)).reshape(xtrain.shape)

xtrain

array([[       0,        1,        2, ...,       20,       21,       22],
       [      23,       24,       25, ...,       43,       44,       45],
       [      46,       47,       48, ...,       66,       67,       68],
       ...,
       [15960597, 15960598, 15960599, ..., 15960617, 15960618, 15960619],
       [15960620, 15960621, 15960622, ..., 15960640, 15960641, 15960642],
       [15960643, 15960644, 15960645, ..., 15960663, 15960664, 15960665]])

In [40]:
xtest = np.arange(0, np.prod(xtest.shape)).reshape(xtest.shape)
xtest

array([[      0,       1,       2, ...,      20,      21,      22],
       [     23,      24,      25, ...,      43,      44,      45],
       [     46,      47,      48, ...,      66,      67,      68],
       ...,
       [6840223, 6840224, 6840225, ..., 6840243, 6840244, 6840245],
       [6840246, 6840247, 6840248, ..., 6840266, 6840267, 6840268],
       [6840269, 6840270, 6840271, ..., 6840289, 6840290, 6840291]])

In [41]:
knn_pred_test_data = knn_model.predict(xtest)
knn_pred_test_data

array(['N', 'N', 'N', ..., 'Y', 'Y', 'Y'], dtype=object)

In [42]:
knn_pred_train_data = knn_model.predict(xtrain)
knn_pred_train_data

array(['N', 'N', 'N', ..., 'Y', 'Y', 'Y'], dtype=object)

In [49]:
from sklearn.metrics import mean_squared_error, mean_absolute_error

In [55]:
accuracy_knn = accuracy_score(ytrain, knn_pred_train_data)
precision_knn = precision_score(ytrain, knn_pred_train_data, pos_label='Y')
recall_knn = recall_score(ytrain, knn_pred_train_data, pos_label='Y')
f1_knn = f1_score(ytrain, knn_pred_train_data, pos_label='Y')
roc_auc_knn = roc_auc_score(ytrain, (knn_pred_train_data == 'Y').astype(int))

# Print the metrics
print("KNN Classifier Metrics for Train Data:")
print("Accuracy:", accuracy_knn)
print("Precision:", precision_knn)
print("Recall:", recall_knn)
print("F1 Score:", f1_knn)
print("ROC AUC Score:", roc_auc_knn)

KNN Classifier Metrics for Train Data:
Accuracy: 0.5003789942098907
Precision: 0.5003768353611351
Recall: 0.9999971200728046
F1 Score: 0.667000906674043
ROC AUC Score: 0.50000432852947


In [57]:
accuracy_knn = accuracy_score(ytest, knn_pred_test_data)
precision_knn = precision_score(ytest, knn_pred_test_data, pos_label='Y')
recall_knn = recall_score(ytest, knn_pred_test_data, pos_label='Y')
f1_knn = f1_score(ytest, knn_pred_test_data, pos_label='Y')
roc_auc_knn = roc_auc_score(ytest, (knn_pred_test_data == 'Y').astype(int))

# Print the metrics
print("KNN Classifier Metrics for Test Data:")
print("Accuracy:", accuracy_knn)
print("Precision:", precision_knn)
print("Recall:", recall_knn)
print("F1 Score:", f1_knn)
print("ROC AUC Score:", roc_auc_knn)

KNN Classifier Metrics for Test Data:
Accuracy: 0.49851380613576146
Precision: 0.49850873742009894
Recall: 0.9999932549559212
F1 Score: 0.6653382878273826
ROC AUC Score: 0.5000100370671577
